mjx/mujoco/mjx/_src/sensor.py

```diff
@@ -600,3 +602,79 @@ def _framelinacc(cvel, cacc, offset):
   )
 
   return d.replace(sensordata=sensordata)
+
+
+def energy_pos(m: Model, d: Data) -> Data:
+    """Calculates position-dependent energy (potential).
+    """
+        
+    # Initialize potential energy
+    energy = jp.array(0.0)
+    if not m.opt.enableflags & EnableBit.ENERGY:
+        return d
+    
+    # Add gravitational potential energy for each body
+    if not m.opt.disableflags & DisableBit.GRAVITY:
+        for i in range(1, m.nbody):  # Skip world body
+            energy -= m.body_mass[i] * jp.dot(m.opt.gravity, d.xipos[i])
+    
+    # Add joint spring potential energy
+    if not m.opt.disableflags & DisableBit.PASSIVE:
+        for i in range(m.njnt):
```

Comment:
I challenge you to figure out how to do this without this for loop iterating over joints.  You should find code elsewhere that does this: look for usage of `scan.flat` in `smooth.py` for some examples.  

`scan` is a bit gnarly and not well super well documented, but should be used wherever possible.